Add --fast flag to update_and_rebuild_libmesh.sh

jwpeterson opened the pull request:
The --fast flag skips the steps of removing the build directory and
configuring libmesh, so we do not allow other configure flags to be
passed in addition to --fast.  We also check that there is a
pre-existing build directory when the user passes --fast, you can't do
--fast the first time you build libmesh.

Refs #6741.

jwpeterson committed:
Add --fast flag to update_and_rebuild_libmesh.sh

The --fast flag skips the steps of removing the build directory and
configuring libmesh, so we do not allow other configure flags to be
passed in addition to --fast.  We also check that there is a
pre-existing build directory when the user passes --fast, you can't do
--fast the first time you build libmesh.

Refs #6741.

jwpeterson commented:
@bwspenc I also verified that `METHODS=opt ./update_and_rebuild_libmesh.sh` works to have libmesh only built in optimized mode.  I think you were referring to making this behavior the default, but I'm not sure we want to do that.  IMO, it's easier to help people debug stuff by just saying "build a debug version of MOOSE" rather than telling them to rerun update_and_rebuild_libmesh.sh with a custom `METHODS` set.  

Now, does the average user need both dbg *and* oprof builds?  I'd say probably not... oprof can give you more human-readable stack traces than opt, but they still don't have the line numbers, dbg mode is the only way to get that last time I checked.  oprof is also used for profiling, but I'm not sure how much of that the average user does.

bwspenc commented:
@jwpeterson I think it's fine to keep the current behavior by default.  One thing, though is that everyone is used to setting `METHOD`, but they'd have to look at the script to know that you have to set `METHODS`.  Should we have it accept both `METHOD` and `METHODS` interchangeably?

jwpeterson commented:
On Tue, Apr 19, 2016 at 3:31 PM, Ben Spencer <notifications@github.com>
wrote:

> @jwpeterson <https://github.com/jwpeterson> I think it's fine to keep the
> current behavior by default. One thing, though is that everyone is used to
> setting METHOD, but they'd have to look at the script to know that you
> have to set METHODS. Should we have it accept both METHOD and METHODS
> interchangeably?
>

METHODS is a libmesh thing, METHOD is a MOOSE thing (that used to be a
libmesh thing).  I don't think libmesh's configure uses METHOD anymore.  We
could translate a user's METHOD into an appropriate METHODS; if it's not
set they will get "opt dbg oprof"?

-- 
John

bwspenc commented:
Since `METHOD` is a MOOSE thing, I think my favorite way to handle that would be to have the script translate that to `METHODS` (only if `METHOD` is specified, but `METHODS` isn't), but it's a little messy, so I'm not really tied to that idea.

jwpeterson commented:
I just discovered with @milljm that PETSc 3.6.4 will require us to pass `--with-metis=PETSc` to the configure script for libmesh.  Doing this *shouldn't* break existing builds, since we know that our PETSc always has a METIS.  I also need to implement the `METHOD`/`METHODS` thing that @bwspenc asked about.

jwpeterson committed:
Allow METHOD=foo ./update_and_rebuild_libmesh.sh

jwpeterson committed:
Pass --with-metis=PETSc when configuring libmesh.

Our recent packages all have a PETSc that comes with SuperLU/MUMPs and
its own METIS/ParMETIS.  libMesh should be set to use this METIS so
there are no conflicts/errors at link time.

permcody commented:
👍

friedmud commented:
@jwpeterson turns out you weren't quite right about `--with-metis=PETSc`: https://groups.google.com/forum/#!topic/moose-users/oUSC6UHrC4c

jwpeterson commented:
Yeah, we jumped the gun making it the default.  Neesds to be taken back out.  This will be handled dynamically in the next libmesh update...

friedmud commented:
@jwpeterson I figured.  No big deal.

We get used to people using our package... sometimes we forget that people really do use MOOSE in other weird configurations :-)
